File: spatial_domain/python/task1_3.py
# coding=UTF-8
# 1.3. Negativo
# (intensidade na saída = 255 - intensidade na entrada)

import utils
import color


def main():
    original_image = utils.load_image('lenna.png')

    grayscale_image = color.rbg2gray(original_image)
    negative_grayscale_image = _invert_image(grayscale_image)
    back_to_grayscale = _invert_image(negative_grayscale_image)

    negative_color_image = _invert_image(original_image)
    back_to_original = _invert_image(negative_color_image)

    utils.display_multiple_images(['Lenna', 'Negative Color Lenna', 'Back to Original'],
                                  [original_image, negative_color_image, back_to_original])

    utils.display_multiple_images(['Grayscale Lenna', 'Negative Grayscale', 'Back to Grayscale'],
                                  [grayscale_image, negative_grayscale_image, back_to_grayscale])

    utils.wait_key_and_destroy_windows()


def _invert_image(image):
    return 255 - image.astype('uint8')


if __name__ == "__main__":
    main()

File: spatial_domain/matlab/task1_3.m
% 1.3. Negativo
% (intensidade na saída = 255 - intensidade na entrada)
function task1_3()
    assetsDir = ['..' filesep 'assets' filesep];
    filePath = strcat(assetsDir, 'lenna.png');
    
    originalImage = imread(filePath);
    negativeImage = getNegative(originalImage);
    backToOriginalImage = getNegative(negativeImage);
        
    % Ploting
    fig2=figure();        
    subplot(1,3,1), imshow(originalImage,[]), title('Original image')
    subplot(1,3,2), imshow(negativeImage,[]), title('Negative image')
    subplot(1,3,3), imshow(backToOriginalImage,[]), title('Back to original image')
       
    grayScale = rgb2gray(originalImage);
    negativeGrayScaleImage = getNegative(grayScale);
    backToGrayScale = getNegative(negativeGrayScaleImage);
    
    fig=figure();        
    subplot(1,3,1), imshow(grayScale,[]), title('Grayscale image')
    subplot(1,3,2), imshow(negativeGrayScaleImage,[]), title('Negative grayscale image')
    subplot(1,3,3), imshow(backToGrayScale,[]), title('Back to grayscale image')
    
    saveas(fig,strcat('output',filesep,mfilename,'_1.png'));
    saveas(fig2,strcat('output',filesep,mfilename,'_2.png'));
    
end

function output = getNegative(input)
    % MatLab shortcut
    % output = imcomplement(input);
    
    output = (255 - input);
end